### DISCUSSION TITLE
Avoid Query Interface when assigning to a variable or passing a parameter

### DISCUSSION THREAD

**Discussion by Greedquest** [twinbasic/twinbasic#1439] [Ideas]
Weird situation. I have a method that obtains an interface `ITypeInfo` @ ObjPtr = 123

However if I call query interface on that object, asking for ITypeInfo, I'll get another object @ObjPtr = 234

This is not COM compliant behaviour, so not really tB's problem I suppose. However that is the situation

```vba
Dim outITypeInfo As ITypeInfo
typelib.getTypeInfo outITypeInfo
Debug.Print objptr(outITypeInfo) '123 - but this is an ITypeInfo instance

Dim anotherITypeInfo As ITypeInfo
Set anotherITypeInfo = outITypeInfo  'Calls IUnk::QI
Debug.Print objptr(anotherITypeInfo ) '234 - a less useful ITypeInfo instance
```

Obviously I can pass pointers around and dereference at the callsite, but then I don't get the benefit of COM reference counting.

It would be good if I could pass an interface like ITypeInfo around and have some guarantees query interface won't be called on it. This is more performant as well. Passing around IUnknown is ok too, but then I'd need a way to cast to ITypeInfo without Query interface. It also doesn't give much type safety.

**Comment by WaynePhillipsEA** [comment]
I recognize that behaviour ;)  When I wrote the C# VBx in-memory ITypeInfo parsing, I had to fight hard to prevent the CLR from calling QueryInterface automatically on those pesky double-implemented ITypeInfos.

Passing an argument ByRef rather than ByVal should avoid QI being called in argument passing.  But for general variable assignments, we don't currently have a way of avoiding it without resorting to raw LongPtrs, since we copy VBx behaviours exactly.

**Reply by Greedquest** [reply]
I thought you might recognise it :)

ByRef is helping a lot I didn't think of that and swapping all my functions form `GetX() As ITypeInfo` to `tryGetX(ByRef outTI As ITypeInfo) As Boolean` helps. Thanks so much.

However there are some places where assignment is unavoidable (e.g. if I cache the typeInfos then want to pop the cache into a variable, QI is called). So it still would be nice to have a mechanism for tracking, in the compiler maybe, "If this variable is declared as ITypeInfo, and then I assign it to another variable with ITypeInfo or supertype IUnknown - e.g. in a function return, QI does not need to be called" that would be typesafe. 

And then a complimentary unsafe `Cast` method to downcast a general container ( IUnknown/Variant ) to a specific subtype ( ITypeInfo ) without calling QI. This may be possible to implement as a Sub at some point with generics + pointer magic already though so may not need tB support.

e.g. (functionality proposal, not syntax)
```vba
Dim outTI As ITypeInfo
typelib.gettypeInfo outTI

NoQuery Return outTI 'does not call QI, raises compiler error if Function return type interface is not same as or `Extend`ed by declared type of variable

...

Dim cache As Dictionary = New Dictionary
Set cache(outTI.Name) = CVar(outTI) 'no QI called

...

Dim tiToGet As String = "foo"
Dim cacheItem As Variant = cache(tiToGet)
Dim unwrapped As IUnknown NoQuery= cacheItem 'assign variant -> IUnknown without QI should always work if vartype = vbObject because COM says so
Return Cast(Of ITypeInfo)(unwrapped) 'IUnknown -> ITypeInfo
```

Feels worthy on grounds of
 - Making the best language for COM that has this kind of flexibility built-in without reverting to pointers - losing refcount safety (and type safety but I imagine strong typed pointers will be popular and get support)
 - Explicit QueryInterface is better than implicit

**Comment by wqweto** [comment]
In VB6 you can use `vbaObjSetAddref` to skip QI on assignment (but call AddRef) like this
```
Private Declare Function vbaObjSetAddref Lib "msvbvm60" Alias "__vbaObjSetAddref" (oDest As Any, ByVal lSrcPtr As Long) As Long

Private Sub Form_Load()
    Dim oObj        As Object
    Dim oCopy       As Object
    
    Set oObj = New Class1
    Call vbaObjSetAddref(oCopy, ObjPtr(oObj))
End Sub
```
I think `vbaObjSetAddref ` API declare is one of the internalized declares on msvbvm60.

The same API function can be used to change interface (change data-type) without QI which might be dangerous unless upcasting (i.e. from ITypeInfo to IDispatch or to IUnknown is safe).

**Reply by Greedquest** [reply]
Ok that's very useful - and for (runtime) type safety you can do something like this:

```vba
Public Class GenericCast(Of TSource, TDest)
	Private Declare PtrSafe Function ShallowCopy Lib "msvbvm60" Alias "__vbaObjSetAddref" (ByRef Dest As Any, ByVal pSourceInterface As LongPtr) As Long
    Public Property Let Cast(ByRef Dest As TDest, ByRef Source As TSource)
        ShallowCopy Dest, ObjPtr(Source)
    End Property
End Class

Private Class myClass
End Class

Private Module Test
	Public Sub PerformCast()
	    Dim x As myClass = New myClass 
	    Dim asIUnk As IUnknown = x
	    Debug.Print ObjPtr(x), ObjPtr(asIUnk) 'different pointers
	    Dim caster As New GenericCast(Of myClass, IUnknown)
	    caster.Cast(asIUnk) = x 'Set asIUnk = x without QI
	    Debug.Print ObjPtr(x), ObjPtr(asIUnk) 'same pointer
	End Sub
End Module
```
